witness/internal/witness/db.go

```diff
@@ -53,8 +54,23 @@ func (d *Database) GetLatest(logPK string) (*Chkpt, error) {
 	return &maxChkpt, nil
 }
 
-// SetCheckpoint writes the checkpoint to the DB for a given logPK.
-func (d *Database) SetCheckpoint(logPK string, c Chkpt) error {
-	_, err := d.db.Exec("INSERT OR IGNORE INTO chkpts (key, size, raw) VALUES (?, ?, ?)", logPK, c.Size, c.Raw)
-	return err
+// SetCheckpoint writes the checkpoint to the DB for a given logPK, assuming
+// that the latest checkpoint is still what the caller thought it was.
+func (d *Database) SetCheckpoint(logPK string, latest, c *Chkpt) error {
+	tx, err := d.db.BeginTx(context.Background(), nil)
```

Comment:
A classic trap! Rule of thumb: the only place that a `context` should be pulled from thin air is in the `main` method. The `ctx` here should be passed into `SetCheckpoint`. The docs for it are pretty useful: https://golang.org/pkg/context/

Reply:
Okay, I've been good now and defined the context only in the test.